flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DefaultSchemaResolver.java

```diff
@@ -0,0 +1,328 @@
+        final ResolvedExpression resolvedExpression;
+        try {
+            resolvedExpression = resolveExpression(inputColumns, unresolvedColumn.getExpression());
+        } catch (Exception e) {
+            throw new ValidationException(
+                    String.format(
+                            "Invalid expression for computed column '%s'.",
+                            unresolvedColumn.getColumnName()),
+                    e);
+        }
+        return Column.computed(unresolvedColumn.getColumnName(), resolvedExpression);
+    }
+
+    private void validateDuplicateColumns(List<Column> columns) {
+        final List<String> names =
+                columns.stream().map(Column::getName).collect(Collectors.toList());
+        final List<String> duplicates =
+                names.stream()
+                        .filter(name -> Collections.frequency(names, name) > 1)
+                        .distinct()
+                        .collect(Collectors.toList());
+        if (duplicates.size() > 0) {
+            throw new ValidationException(
+                    String.format(
+                            "Schema must not contain duplicate column names. Found duplicates: %s",
+                            duplicates));
+        }
+    }
+
+    private List<WatermarkSpec> resolveWatermarkSpecs(
+            List<UnresolvedWatermarkSpec> unresolvedWatermarkSpecs, List<Column> inputColumns) {
+        if (unresolvedWatermarkSpecs.size() == 0) {
+            return Collections.emptyList();
+        }
+        if (unresolvedWatermarkSpecs.size() > 1) {
+            throw new ValidationException("Multiple watermark definitions are not supported yet.");
+        }
+        final UnresolvedWatermarkSpec watermarkSpec = unresolvedWatermarkSpecs.get(0);
+
+        // validate time attribute
+        final String timeColumn = watermarkSpec.getColumnName();
+        validateTimeColumn(timeColumn, inputColumns);
+
+        // resolve watermark expression
+        final ResolvedExpression watermarkExpression;
+        try {
+            watermarkExpression =
+                    resolveExpression(inputColumns, watermarkSpec.getWatermarkExpression());
+        } catch (Exception e) {
+            throw new ValidationException(
+                    String.format(
+                            "Invalid expression for watermark '%s'.", watermarkSpec.toString()),
+                    e);
+        }
+        validateWatermarkExpression(watermarkExpression.getOutputDataType().getLogicalType());
+
+        return Collections.singletonList(
+                new WatermarkSpec(watermarkSpec.getColumnName(), watermarkExpression));
+    }
+
+    private void validateTimeColumn(String columnName, List<Column> columns) {
+        final Optional<Column> timeColumn =
+                columns.stream().filter(c -> c.getName().equals(columnName)).findFirst();
+        if (!timeColumn.isPresent()) {
```

Comment:
nit: we can avoid the `Optional` here:

```
final Column timeColumn = columns.stream().filter(c -> c.getName().equals(columnName))
  .findFirst()
  .orElseThrow(() -> new ValidationException(…));
```

Reply:
I did this for code readability, using `orElseThrow` will blow up the formatting